### DISCUSSION TITLE
c_cpp_properties.json vscode variables

### DISCUSSION THREAD

**Discussion by LostTime76** [microsoft/vscode-cpptools#12997] [Q&A] [answered]
I am trying to use vscode variables within c_cpp_properties.json, and I do not understand the resolution strategy. All seems to be working like I expect; however, there is a contradiction.

1. If I put ${fileDirname} within a tasks.json file and run the task, it resolves to the directory name of my currently opened editor file. 
2. If I put ${fileDirname} within a c_cpp_properties.json includes configuration, it **seems** to resolve to the directory of c_cpp_properties.json regardless of any opened editor file (this is absolutely the desired behavior)

I need / want a means to substitute a vscode variable that **_selects the current file_**. **_Not the currently open file_** within the editor. Vscode should provide such variables, but the variable reference only states the ${file_xx} variables resolving to the currently open editor file.

What is the concrete implementation here as there is clearly a contradiction. I do not want to rely on something within c_cpp_properties.json if its just a fluke. If it is a fluke, I am going to put in a feature request at light speed, because it would be quite ridiculous to only provide ${file_xx} variable resolution with regards to the  currently open editor. I (we) need a way to have variables resolve to the current file.

How am I testing?
I am creating two c_cpp_properties.json configurations that have two different include paths (using ${file_xx} variables) depending on the selected configuration. In a c source file, I am using a define that will be defined based on the header file with the same name within the included path of the selected configuration. The intellisense picks up the correct define when I switch between the configurations. If I change the path of the c source file, the correct define is still picked up. The c source file is using <> header inclusion to treat the include path as a system include. 

This indicates that ${file_xx} used within c_cpp_properties.json is resolving to the directory of c_cpp_properties.json regardless of any open editor file. This is the desired behavior but is a contradiction to how its resolved within tasks.json and also what the documentation states.

**Comment by bobbrow** [comment]
Hi @LostTime76,

Thank you for your question. ${fileDirname} is not a defined "special" variable for `c_cpp_properties.json`. What I suspect that you are seeing is that we interpreted it as an environment variable and since it was not defined in the environment we replaced it with an empty string. After that we interpreted the remainder of your path as a relative path and prepended the value of ${workspaceFolder} to it when constructing your include path.

The variable you seem to be describing that you want is `${workspaceFolder}` which is a supported "special" variable in `c_cpp_properties.json`.

Could you point to the documentation that you are referencing?  I would guess you were referring to the VS Code documentation for Tasks which lists many variables. However, that is part of the core editor functionality and `c_cpp_properties.json` is defined by our extension so that documentation does not fully apply here. We have added support for many of the variables defined in the Tasks documentation to try to reduce confusion, but we did not support all of them due to the fact that we need a stable include path, not one that can change with the focused file.

**Reply by sean-mcmanus** [reply]
@bobbrow The user wants some `${currentlyActiveFile}`, which is not something we support, but `${workspaceFolder}` might/should work with a multi-root workspace folder and switching between files in the different workspace folders.

**Reply by bobbrow** [reply]
That was not what I understood from the question.

> This indicates that **${file_xx} used within c_cpp_properties.json is resolving to the directory of c_cpp_properties.json** regardless of any open editor file. **This is the desired behavior** but is a contradiction to how its resolved within tasks.json and also what the documentation states.

**Reply by sean-mcmanus** [reply]
@bobbrow Ah, okay, you are right.

**Comment by LostTime76** [comment]
The documentation I am referencing is here:
https://code.visualstudio.com/docs/editor/variables-reference

Yes, I would like a 'stable' variable that does not change with the focused file. I would like to reference a path relative to the current file. In general, the user will be opening the workspace in the root folder, which will have .vscode folder, so ${workspaceFolder} would potentially work. However, it would be nice if there were variables that could just reference the current file / directory, so relative paths could be constructed without worrying about that. 

Maybe its not a huge deal... and I just have to tell people that the workspace needs to be opened on a specific folder containing the .vscode directory for everything to work. That way we can just use ${workspaceFolder}.

**Reply by bobbrow** [reply]
I still struggle to understand what you would do with such a variable. The settings within `c_cpp_properties.json` describe configurations for the entire workspace, so there isn't a concept of "current file" besides the `c_cpp_properties.json` file itself.  It seems that `${workspaceFolder}` is what you want, but I still feel like I'm missing something.


> Maybe its not a huge deal... and I just have to tell people that the workspace needs to be opened on a specific folder containing the .vscode directory for everything to work. That way we can just use ${workspaceFolder}.

If you are using `c_cpp_properties.json` to define your configurations, ^this is actually the contract.  We don't go searching around in the workspace for multiple `c_cpp_properties.json` files. We only look at `${workspaceFolder}\.vscode\c_cpp_properties.json`

**Reply by LostTime76** [reply, marked as answer]
I think I talked myself out of the need for what I was looking for.. The contract is that the workspace must be opened with the proper path, otherwise it won't work.

My use case was perhaps generating / consuming the c_cpp_properties.json file using a custom tool and the tool being able to resolve a subset of the vscode variables. Since the tool would have no concept of ${worskspaceFolder}, a variable just resolving to the current path of the file directly would be better. I guess the tool could just resolve ${workspaceFolder} as the directory above the directory the c_cpp_properties.json is within as a special case.. because in order for even vscode to correctly work, the correct folder needs to be opened, which would always be the directory above c_cpp_properties.json. Besides the point... The tool could also just generate absolute paths too...

This can be closed.

Thanks

**Reply by bobbrow** [reply]
Thank you for the explanation. I think I understand a little better now. We will mark this discussion as answered.
